### ISSUE TITLE
Semantic Search: Try Michael's model for semantic search

### ISSUE THREAD

**Issue by rachlllg** [freelawproject/courtlistener#5166]
Michael Bommarito shared some models with us that we could potentially use for semantic search, this issue is to document the work I need to do to run experiments on his models

**Comment by rachlllg** [comment]
Since Michael's models are pretrained encoder models, to use for semantic search, they will need to be finetuned using query-document pairs or triples. So I had to create the dataset, and since the dataset can be used for finetuning, I went ahead and finetuned numerous encoder models on this dataset to evaluate the performance across the board.

#### Data Preparation
1. Used the train and test datasets generated above, identified `docket_numbers` that are in both train and test and put those records as `validation`. This ensures no opinion from the same docket or the same cluster appear in both the train and the test set (no data leakage).
   - This created 390 unique opinions in train, 113 unique opinions in val, and 450 unique opinions in test (the test set remains the same as above)
2. Given opinions of very short length would not provide meaningful retrieval, I removed any records that have fewer than 50 words in length (~100 tokens)
   - This created 315 unique opinions in train, 93 unique opinions in val, 362 unique opinions in test
3. Chunk the train and val opinions to chunks with at most 512 tokens and 8192 tokens, chunked with sentence boundary and 2-sentence overlap between chunks
   - For the set with 512 context window, this created 2828 unique chunks in train and 489 unique chunks in val
   - For the set with 8192 context window, this created 351 unique chunks in train, 95 unique chunks in val
4. Used GPT-4o with a slightly different prompt from above to generate relevant and irrelevant questions for each chunk
   - I selected GPT-4o here (vs GPT-4o mini above) for its superior intelligence since the data here will be used for training
   - I used a slightly different prompt from above to further ensure the robustness in the evaluation (no data leakage)
5. For final evaluation, I will use the same test set as used in above evaluations to ensure the results are comparable

#### Model finetuning
1. I selected below models for finetuning with the chunk-relevant-irrelevant triples generated above
   - bert-base-cased
   - roberta-base
   - microsoft/mpnet-base
   - answerdotai/ModernBERT-base
   - alea-institute/kl3m-doc-pico-001
   - alea-institute/kl3m-doc-nano-001
   - alea-institute/kl3m-doc-micro-uncased-001
   - alea-institute/kl3m-doc-small-uncased-001
   - nomic-ai/modernbert-embed-base
2. With 512 context window, all models were trained for only 1 epoch and all hyperparameters used for finetuning were the same except for adjusting for batch_size to accomodate the GPU RAM limitation (I used the T4 GPU with 15GB GPU RAM provided for free by Google Colab).
3. I was not able to train the ModernBERT based models with the 8192 long context dataset due to GPU OOM error, and flash attention is not supported on the version of GPU used for this task.

#### Model Evaluation
To ensure the results are comparable to the experiments already conducted above, I used the same 450 test datapoints to evaluate the model performance. It should be noted that in production, short opinions will be removed which should lead to a slight performance increase.

Below is the summary of results. We still see the `nomic-ai/modernbert-embed-base` has the best performance and we do see a sizable improvement from finetuning. In fact, the finetuned model with 512 chunk size now surpasses the original model with 8192 chunk size. Finetuning the model with 8192 chunk size data should further improve the performance of the 8192 chunk size inference.

<img width="834" alt="Image" src="https://github.com/user-attachments/assets/6a7b4272-5890-4c70-aeb4-d36404ccb708" />

Given this is a synthetic dataset and we do not have the golden dataset yet (which I will work with SME to create), and finetuning will take additional time and resource, at this stage, I recommend we stay with the model we already selected, unless @mlissner prefer otherwise based on the above results?

**Comment by mlissner** [comment]
> identified docket_numbers that are in both train and test and put those records as validation. 

Note that docket numbers are only unique *within* a court, but that they're repeated *across* courts. So if you have two docket numbers that are the same, it's likely the cases are from different courts.

> I recommend we stay with the model we already selected, unless @mlissner prefer otherwise based on the above results?

Your recommendation sounds good to me, but if you're getting enhanced results with this simple finetuned dataset, should we use this basic finetuning? 

> I used the T4 GPU with 15GB GPU RAM provided for free by Google Colab

If you can share your work, that'd be really good for future replication. Maybe link it here at least?

**Comment by rachlllg** [comment]
> Note that docket numbers are only unique within a court, but that they're repeated across courts. So if you have two docket numbers that are the same, it's likely the cases are from different courts.

Ohhh I did not realize this, thank you! This means my data was fine to begin with, but this still works.

> Your recommendation sounds good to me, but if you're getting enhanced results with this simple finetuned dataset, should we use this basic finetuning?

If we want to use the finetuned one, I'd love to finetune on the 8192 chunk size too to see if it improves, since we will embed with the larger chunk size. Is it okay with you if I spin up a GPU instance in EC2 to do this finetuning? I already have the data & the code, just need a GPU instance with enough memory for the large chunk size.

> If you can share your work, that'd be really good for future replication. Maybe link it here at least?

Ah, thanks for reminding me. 

Public on my personal repo: https://github.com/rachlllg/FLP_Semantic_Search
Also made a copy in the FLP repo: https://github.com/freelawproject/ai-experiments/tree/FLP_Semantic_Search

**Comment by mlissner** [comment]
> Is it okay with you if I spin up a GPU instance in EC2 to do this finetuning? 

Sure, that's fine, I think. How much money and time will it be? Please put in an expense request for your AWS costs.

**Comment by rachlllg** [comment]
> How much money and time will it be?

I think a T4 GPU with high RAM would suffice, I will start with the 32GB one ($0.75/hr) and if that doesn't work, move up to a higher RAM version, I highly doubt I will need one with more than 128GB RAM, so at most a couple dollars an hour. I shouldn't need it for more than an hour or two, I already have the code, just need to run it.

https://aws.amazon.com/ec2/instance-types/g4/

**Comment by mlissner** [comment]
Sounds great!

**Comment by rachlllg** [comment]
Turns out the T4 GPUs do not have enough GPU RAM so I used `g5.2xlarge` which is an A10 GPU with 24GB GPU RAM. I could only fit 2 batches (with 8192 context window) into the GPU RAM (vs 16 batches with the 512 context window) so I reduced the learning rate and trained for 2 epochs instead of 1.

The instance was up for ~3-4 hours (@$1.212/hr) of which ~2 hours was for setting up the env, GPU driver, and cuda compatible packages. I experimented with the batch size to find the largest batch size allowed and tweaked the learning rate and epochs to reduce overfitting, and ran eval.

https://aws.amazon.com/ec2/instance-types/g5/
https://aws.amazon.com/ec2/instance-types/

Here is the eval results. 

- The `Free-Law-Project/modernbert-embed-base_finetune_xxx` models are what we finetuned on top of the `nomic-ai/modernbert-embed-base` model. 
- In evaluating the models using the same context window size matching the training size (if the model was finetuned with 512 chunk size, the eval data was also chunked to 512 chunk size, same for 8192 chunk size). 
   - We see that the `Free-Law-Project/modernbert-embed-base_finetune_512` model have the best overall performance. 
   - We should have expected the `Free-Law-Project/modernbert-embed-base_finetune_8192` model to have the best performance, but it is important to note that due to the chunk size difference, the number of training datapoints for the 512 chunk size model is substantially more than that for the 8192 chunk size model (ie, an opinion with 2000 tokens would have produced ~4 chunks with 512 chunk size vs 1 chunk for 8192 chunk size). 
   - We still see a slight improvement in the 8192 chunk size model after finetuning, but more training datapoints is needed to further boost the performance.
- To evaluate the best model overall, I also ran eval with the `Free-Law-Project/modernbert-embed-base_finetune_512` model on the eval data chunked to 8192 chunk size. The result is comparable to that of the `Free-Law-Project/modernbert-embed-base_finetune_512` model on the eval data chunked to 512 chunk size: marginally higher Hit-Rate and marginally lower MRR.
- I want to emphasize that the evals are run on an opinion level since the search results are presented to users on an opinion level. So each opinion is chunked to multiple chunks, the retrieval is done on the chunk level, the eval is based on "whether the chunk from an opinion appears in top-k".

![Image](https://github.com/user-attachments/assets/dfae7568-a1d7-4465-9280-104dd55517e3)

Based on this, I recommend we use the `Free-Law-Project/modernbert-embed-base_finetune_512` model, with opinions chunked to either 512 chunk size or 8192 chunk size. Below is a pro vs con for consideration.

| | 512 chunk size | 8192 chunk size |
|----------|----------|----------|
| Pros    | - We may leverage this to highlight the chunk within the opinion (that the retrieval returned) to present to the user  <br> - Bommarito noted their experiments showed "page" level retrieval may have better performance from their experiments | - More contextually rich embeddings <br> - Most opinions are embedded to one 768-sized embedding, reducing storage and memory footprint in retrieval at scale |
| Cons    | - Most opinions are embedded to multiple 768-sized embeddings, substantially increasing the storage and memory footprint in retrieval at scale | - May be difficult to highlight since most opinions are represented with one chunk(one embedding) <br> - The finetuned model is finetuned with 512 chunk size, it would be better to finetune with more longer context data to match how we use it at inference  |

The GitHub repos has been updated, and the dataset + models are posted to [HuggingFace](https://huggingface.co/collections/Free-Law-Project/semantic-search-67c76644a9d04a1126a78908)

**Comment by mlissner** [comment]
Hm, sounds like the only downside to 512 is performance/storage/memory consumption, but that we don't know the scale of that impact. Meanwhile, if we go with the 512 chunk size, we get better highlighting, which is a fairly important feature for users. 

Balancing that, I think we should go with the 512, and then if performance is an issue, we re-evaluate.
